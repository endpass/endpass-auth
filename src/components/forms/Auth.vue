<template>
  <form data-test="auth-form" @submit.prevent="handleSubmit">
    <form-field v-if="isServerMode">
      <server-mode-select v-model="serverMode" @confirm="handleSubmit" />
    </form-field>
    <template v-if="isDefaultMode">
      <form-field>
        <message data-test="form-message">
          Log in to your Endpass account to access site actions
        </message>
      </form-field>
      <form-field v-if="error">
        <message :error="true" data-test="error-message">{{ error }}</message>
      </form-field>
      <form-field>
        <div class="auth__fields-as-line">
          <v-input
            v-model="email"
            :invalid="!isEmailValid"
            :autofocus="true"
            name="email"
            type="email"
            placeholder="Enter your email..."
            data-test="email-input"
          />
          <v-button
            :disabled="!isFormValid"
            :submit="true"
            type="primary"
            data-test="submit-button"
            >{{ primaryButtonLabel }}</v-button
          >
        </div>
      </form-field>
      <form-controls>
        <google-auth-button @error="handleOauthError" />
      </form-controls>
      <form-controls>
        <git-auth-button @error="handleOauthError" />
      </form-controls>
      <form-controls>
        <v-checkbox v-model="isTermsAccepted">
          I accept the
          <a href="https://endpass.com/terms/" target="_blank"
            >Terms of Service</a
          >
          and
          <a href="https://endpass.com/privacy/" target="_blank"
            >Privacy Policy</a
          >.
        </v-checkbox>
      </form-controls>
    </template>
  </form>
</template>

<script>
import Vue from 'vue';

import VCheckbox from '@endpass/ui/dist/components/VCheckbox';
import VFrame from '@/components/common/VFrame.vue';
import VInput from '@/components/common/VInput.vue';
import VButton from '@/components/common/VButton.vue';
import GoogleAuthButton from '@/components/common/GoogleAuthButton.vue';
import GitAuthButton from '@/components/common/GitAuthButton.vue';
import Message from '@/components/common/Message.vue';
import FormField from '@/components/common/FormField.vue';
import FormControls from '@/components/common/FormControls.vue';
import ServerModeSelect from '@/components/common/ServerModeSelect';
import { IDENTITY_MODE } from '@/constants';

Vue.component(VCheckbox);

export default {
  name: 'AuthForm',

  props: {
    inited: {
      type: Boolean,
      default: false,
    },

    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    isServerMode: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    isTermsAccepted: true,
    email: '',
    serverMode: {
      type: IDENTITY_MODE.DEFAULT,
      serverUrl: undefined,
    },
  }),

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Log in' : 'Loading...';
    },

    isEmailValid() {
      return /^[a-zA-Z._\-\+0-9]+@[a-z0-9]+\.[a-z]{2,}$/g.test(this.email);
    },

    isDefaultMode() {
      return this.serverMode.type === IDENTITY_MODE.DEFAULT;
    },

    isLocalMode() {
      return this.serverMode.type === IDENTITY_MODE.LOCAL;
    },

    isCustomMode() {
      return this.serverMode.type === IDENTITY_MODE.CUSTOM;
    },

    isFormValid() {
      const {
        isDefaultMode,
        isLocalMode,
        isCustomMode,
        isEmailValid,
        isTermsAccepted,
        loading,
      } = this;
      const isDefaultValid = isDefaultMode && isEmailValid && isTermsAccepted;

      return (isDefaultValid || isCustomMode || isLocalMode) && !loading;
    },
  },

  methods: {
    handleSubmit() {
      if (!this.isFormValid) return;

      const { email, serverMode } = this;

      this.$emit('submit', { email, serverMode });
    },

    handleOauthError(err) {
      this.$emit('error', err);
    },
  },
  components: {
    VCheckbox,
    VFrame,
    VButton,
    VInput,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormField,
    FormControls,
    ServerModeSelect,
  },
};
</script>
<style>
.auth__fields-as-line {
  display: flex;
}
</style>
