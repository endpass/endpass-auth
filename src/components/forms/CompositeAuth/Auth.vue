<template>
  <form
    data-test="auth-form"
    @submit.prevent="handleSubmit"
  >
    <form-field v-if="isServerMode && !isPublic">
      <server-mode-select
        v-model="serverMode"
        @confirm="handleSubmit"
      />
    </form-field>
    <template v-if="isDefaultMode">
      <form-field>
        <message
          class="v-modal-card-title"
          data-test="form-message"
        >
          Log in to your Endpass account to access site actions
        </message>
      </form-field>
      <form-field v-if="error">
        <message
          :error="true"
          data-test="error-message"
        >
          {{ error }}
        </message>
      </form-field>
      <form-field>
        <v-input
          v-model="email"
          v-validate="'required|email'"
          data-vv-as="email"
          data-vv-name="email"
          :error="errors.first('email')"
          autofocus="true"
          name="email"
          type="email"
          placeholder="Enter your email..."
          data-test="email-input"
        />
      </form-field>
      <form-field>
        <v-button
          :disabled="!isFormValid"
          size="big"
          data-test="submit-button"
        >
          {{ primaryButtonLabel }}
        </v-button>
      </form-field>
      <v-divider>or sign in with</v-divider>
      <form-controls>
        <google-auth-button
          type="button"
          @submit="handleSocialSubmit"
          @error="handleOauthError"
        />
        <div class="social-buttons-gap" />
        <git-auth-button
          type="button"
          @submit="handleSocialSubmit"
          @error="handleOauthError"
        />
      </form-controls>
      <v-divider />
      <form-controls>
        <v-checkbox v-model="isTermsAccepted">
          I accept the
          <a
            href="https://endpass.com/terms/"
            target="_blank"
          >Terms of Service</a>
          and
          <a
            href="https://endpass.com/privacy/"
            target="_blank"
          >Privacy Policy</a>.
        </v-checkbox>
      </form-controls>
    </template>
  </form>
</template>

<script>
import VCheckbox from '@endpass/ui/kit/VCheckbox';
import VInput from '@endpass/ui/kit/VInput';
import VButton from '@endpass/ui/kit/VButton';
import VDivider from '@endpass/ui/kit/VDivider';
import GoogleAuthButton from '@/components/common/GoogleAuthButton.vue';
import GitAuthButton from '@/components/common/GitAuthButton.vue';
import Message from '@/components/common/Message.vue';
import FormField from '@/components/common/FormField.vue';
import FormControls from '@/components/common/FormControls.vue';
import ServerModeSelect from '@/components/common/ServerModeSelect';
import { IDENTITY_MODE } from '@/constants';
import formMixin from '@/mixins/form';

export default {
  name: 'AuthForm',

  props: {
    isInited: {
      type: Boolean,
      default: false,
    },

    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    isPublic: {
      type: Boolean,
      default: false,
    },

    isServerMode: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    isTermsAccepted: true,
    email: '',
    serverMode: {
      type: IDENTITY_MODE.DEFAULT,
      serverUrl: undefined,
    },
  }),

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Login' : 'Loading...';
    },

    emailErrorMessage() {
      return this.isEmailValid ? null : 'Invalid email';
    },

    isEmailValid() {
      // eslint-disable-next-line
      return /^[a-zA-Z._\-\\+0-9]+@[a-z0-9]+\.[a-z]{2,}$/g.test(this.email);
    },

    isDefaultMode() {
      return this.serverMode.type === IDENTITY_MODE.DEFAULT;
    },

    isLocalMode() {
      return this.serverMode.type === IDENTITY_MODE.LOCAL;
    },

    isCustomMode() {
      return this.serverMode.type === IDENTITY_MODE.CUSTOM;
    },

    isFormValid() {
      const {
        isDefaultMode,
        isLocalMode,
        isCustomMode,
        isEmailValid,
        isTermsAccepted,
        loading,
      } = this;
      const isDefaultValid = isDefaultMode && isEmailValid && isTermsAccepted;

      return (isDefaultValid || isCustomMode || isLocalMode) && !loading;
    },
  },

  methods: {
    handleSubmit() {
      if (!this.isFormValid) return;

      const { email, serverMode } = this;

      this.$emit('submit', { email, serverMode });
    },

    handleSocialSubmit() {
      this.$emit('socialSubmit');
    },

    handleOauthError(err) {
      this.$emit('error', err);
    },
  },
  mixins: [formMixin],
  components: {
    VCheckbox,
    VButton,
    VInput,
    VDivider,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormField,
    FormControls,
    ServerModeSelect,
  },
};
</script>

<style lang="postcss">
.social-buttons-gap {
  width: 30px;
}
</style>
