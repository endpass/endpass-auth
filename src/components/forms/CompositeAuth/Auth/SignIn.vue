<template>
  <form
    data-test="auth-form"
    @submit.prevent="onSignIn"
  >
    <form-item v-if="isServerMode && !isPublic">
      <v-title>
        <span v-html="$t('components.serverModeSelect.identityServer')" />
      </v-title>
      <server-mode-select
        v-model="serverMode"
        @confirm="onSubmit"
      />
      <v-spacer :height="4" />
      <v-divider />
    </form-item>
    <template v-if="isDefaultMode">
      <v-title>
        <span v-html="$t('components.auth.loginToContinue')" />
      </v-title>
      <form-item v-if="error">
        <message
          :error="true"
          data-test="error-message"
        >
          {{ error }}
        </message>
      </form-item>
      <form-item>
        <v-input
          v-model="email"
          v-validate="'required|email'"
          data-vv-as="email"
          data-vv-name="email"
          :error="errors.first('email')"
          name="email"
          type="email"
          :placeholder="$t('components.auth.enterEmail')"
          data-test="email-input"
        />
      </form-item>
      <v-button
        :disabled="!isSubmitEnable"
        :is-loading="isLoading"
        size="big"
        data-test="submit-button-auth"
      >
        {{ $t('global.continue') }}
      </v-button>
      <v-spacer :height="3" />
      <v-divider>{{ $t('components.auth.orSignInWith') }}</v-divider>
      <form-row>
        <form-controls>
          <google-auth-button
            type="button"
            @submit="onSocialSubmit"
            @error="onOauthError"
          />
          <git-auth-button
            type="button"
            @submit="onSocialSubmit"
            @error="onOauthError"
          />
        </form-controls>
      </form-row>
      <v-divider />
      <form-row class="v-text-center v-fw-b">
        {{ $t('components.auth.dontHaveAccount') }}&nbsp;
        <v-link
          href="#"
          data-test="switch-to-sign-up"
          @click.prevent="onSwitch"
        >
          {{ $t('components.auth.signUp') }}
        </v-link>
      </form-row>
    </template>
  </form>
</template>

<script>
import VInput from '@endpass/ui/kit/VInput';
import VButton from '@endpass/ui/kit/VButton';
import VDivider from '@endpass/ui/kit/VDivider';
import VLink from '@endpass/ui/kit/VLink';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import VSpacer from '@/components/common/VSpacer';
import ServerModeSelect from '@/components/common/ServerModeSelect';
import GoogleAuthButton from '@/components/common/GoogleAuthButton';
import GitAuthButton from '@/components/common/GitAuthButton';
import Message from '@/components/common/Message';
import { IDENTITY_MODE } from '@/constants';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';
import { coreStore, authStore } from '@/store';
import FormControls from '@/components/common/FormControls';

export default {
  name: 'SignInForm',

  coreStore,
  authStore,

  props: {
    isPublic: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    isLoading: false,
    email: '',
    error: '',
    serverMode: {
      type: IDENTITY_MODE.DEFAULT,
      serverUrl: undefined,
    },
  }),

  computed: {
    isServerMode() {
      return this.$options.coreStore.isServerMode;
    },

    isDefaultMode() {
      return this.serverMode.type === IDENTITY_MODE.DEFAULT;
    },

    isLocalMode() {
      return this.serverMode.type === IDENTITY_MODE.LOCAL;
    },

    isCustomMode() {
      return this.serverMode.type === IDENTITY_MODE.CUSTOM;
    },

    isSubmitEnable() {
      const {
        isDefaultMode,
        isLocalMode,
        isCustomMode,
        isFormValid,
        isLoading,
      } = this;
      const isDefaultValid = isDefaultMode && isFormValid;

      return (isDefaultValid || isCustomMode || isLocalMode) && !isLoading;
    },
  },

  watch: {
    email() {
      this.error = null;
    },
  },

  methods: {
    onSwitch() {
      this.$emit('sign-in');
    },

    onSubmit() {
      if (!this.isSubmitEnable) {
        return;
      }
      this.handleSubmit();
    },

    handleSubmit() {
      const { email, serverMode } = this;

      this.$emit('submit', { email, serverMode });
    },

    async onSignIn() {
      if (!this.isSubmitEnable) {
        return;
      }

      try {
        const { email, serverMode } = this;

        if (serverMode.type !== IDENTITY_MODE.DEFAULT) {
          this.handleSubmit();
          return;
        }

        this.isLoading = true;
        await this.$options.authStore.loadAuthChallenge({ email });

        this.handleSubmit();
      } catch (error) {
        this.error = error;
      } finally {
        this.isLoading = false;
      }
    },

    async onSocialSubmit() {
      this.isLoading = true;
      await this.$options.authStore.waitLogin();
      this.$emit('social');
      this.isLoading = false;
    },

    onOauthError() {
      this.error = this.$i18n.t('components.compositeAuth.authFailed');
    },
  },
  mixins: [formMixin],
  components: {
    FormControls,
    VLink,
    VTitle,
    VButton,
    VInput,
    VSpacer,
    VDivider,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormItem,
    FormRow,
    ServerModeSelect,
  },
};
</script>
