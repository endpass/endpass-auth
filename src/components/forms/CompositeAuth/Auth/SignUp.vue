<template>
  <form
    data-test="auth-form"
    @submit.prevent="onSubmit"
  >
    <v-title>
      <span v-html="$t('components.auth.signUpToContinue')" />
    </v-title>
    <form-item v-if="error">
      <message
        :error="true"
        data-test="error-message"
      >
        {{ error }}
      </message>
    </form-item>
    <form-item>
      <v-input
        v-model="email"
        v-validate="'required|email'"
        data-vv-as="email"
        data-vv-name="email"
        :error="errors.first('email')"
        name="email"
        type="email"
        :placeholder="$t('components.auth.enterEmail')"
        data-test="email-input"
      />
    </form-item>
    <form-item>
      <v-input
        v-model="password"
        v-validate="'required|min:8'"
        data-vv-as="password"
        data-vv-name="password"
        :error="errors.first('password') || error"
        name="password"
        type="password"
        required
        :placeholder="$t('components.auth.password')"
        data-test="password-input"
      />
    </form-item>
    <form-item class="v-mb-24">
      <v-input
        v-model="confirmPassword"
        v-validate="'required|min:8'"
        data-vv-as="password"
        data-vv-name="confirmPassword"
        :error="errors.first('confirmPassword') || error"
        name="confirmPassword"
        type="password"
        :placeholder="$t('components.auth.confirmPassword')"
        data-test="repeat-password-input"
      />
    </form-item>
    <v-button
      :disabled="!isSubmitEnable"
      size="big"
      data-test="submit-button-auth"
    >
      {{ primaryButtonLabel }}
    </v-button>
    <v-spacer :height="3" />
    <v-divider>{{ $t('components.auth.orRegisterWith') }}</v-divider>
    <form-item>
      <form-controls>
        <google-auth-button
          type="button"
          @submit="onSocial"
          @error="onOauthError"
        />
        <git-auth-button
          type="button"
          @submit="onSocial"
          @error="onOauthError"
        />
      </form-controls>
    </form-item>
    <form-row class="v-fs-12 v-lh-1-4 v-text-center">
      <div>
        {{ $t('components.auth.iAccept') }}
        <v-link
          href="https://endpass.com/terms/"
          target="_blank"
          is-underline
        >
          <span v-html="$t('components.auth.termsOfService')" />
        </v-link>
        {{ $t('components.auth.and') }}
        <v-link
          href="https://endpass.com/privacy/"
          target="_blank"
          is-underline
        >
          <span v-html="$t('components.auth.privacyPolicy')" />
        </v-link>
      </div>
    </form-row>
    <v-divider />
    <form-row class="v-text-center v-fw-b">
      {{ $t('components.auth.haveAccount') }}&nbsp;
      <v-link
        href="#"
        data-test="switch-to-sign-ip"
        @click.prevent="onSwitch"
      >
        {{ $t('components.auth.signIn') }}
      </v-link>
    </form-row>
  </form>
</template>

<script>
import VInput from '@endpass/ui/kit/VInput';
import VButton from '@endpass/ui/kit/VButton';
import VDivider from '@endpass/ui/kit/VDivider';
import VLink from '@endpass/ui/kit/VLink';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import VSpacer from '@/components/common/VSpacer';
import GoogleAuthButton from '@/components/common/GoogleAuthButton';
import GitAuthButton from '@/components/common/GitAuthButton';
import Message from '@/components/common/Message';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';
import FormControls from '@/components/common/FormControls';
import { authStore } from '@/store';

export default {
  name: 'SignUpForm',

  authStore,

  data: () => ({
    error: '',
    email: '',
    password: '',
    confirmPassword: '',
    isLoading: false,
  }),

  computed: {
    primaryButtonLabel() {
      return !this.isLoading
        ? this.$i18n.t('components.auth.signUp')
        : this.$i18n.t('global.loading');
    },

    isPasswordEqual() {
      return this.confirmPassword === this.password;
    },

    isSubmitEnable() {
      const { isFormValid, isLoading, isPasswordEqual } = this;

      return isPasswordEqual && isFormValid && !isLoading;
    },
  },

  watch: {
    email() {
      this.error = '';
    },
    password() {
      this.error = '';
    },
    confirmPassword() {
      this.error = '';
    },
  },

  methods: {
    onSwitch() {
      this.$emit('sign-up');
    },

    async onSubmit() {
      if (!this.isSubmitEnable) {
        return;
      }
      const { email, password } = this;
      this.$emit('submit', { email, password, isSignUp: true });
    },

    async onSocial() {
      await this.$options.authStore.waitLogin();
      this.$emit('social');
    },

    onOauthError() {
      this.error = this.$i18n.t('components.compositeAuth.authFailed');
    },
  },
  mixins: [formMixin],
  components: {
    FormControls,
    VLink,
    VTitle,
    VButton,
    VInput,
    VSpacer,
    VDivider,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormItem,
    FormRow,
  },
};
</script>
