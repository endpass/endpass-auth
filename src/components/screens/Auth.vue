<template>
  <screen @close="handleAuthCancel">
    <v-frame
      :loading="!isInited"
      :closable="isDialog"
      @close="handleAuthCancel"
    >
      <composite-auth-form
        v-if="activeForm === FORMS.AUTH"
        :closable="isDialog"
        @authorize="handleAuthorize"
      />
      <create-wallet-form
        v-else-if="activeForm === FORMS.CREATE_WALLET"
        @request="handleAccountRequest"
      />
    </v-frame>
  </screen>
</template>

<script>
import { mapState, mapGetters, mapActions } from 'vuex';
import Screen from '@/components/common/Screen';
import VFrame from '@/components/common/VFrame';
import CompositeAuthForm from '@/components/forms/CompositeAuth';
import CreateWalletForm from '@/components/forms/CreateWallet';

const FORMS = {
  AUTH: 'AUTH',
  CREATE_WALLET: 'CREATE_WALLET',
};

export default {
  name: 'Auth',

  data: () => ({
    isCheckingAccount: false,
    FORMS,
    activeForm: FORMS.AUTH,
  }),

  computed: {
    ...mapState({
      isInited: state => state.core.isInited,
      showCreateAccount: state => state.core.showCreateAccount,
    }),

    ...mapGetters(['isDialog']),
  },

  methods: {
    ...mapActions([
      'confirmAuth',
      'cancelAuth',
      'dialogClose',
      'openCreateAccountPage',
      'checkAccountExists',
      'waitAccountCreate',
    ]),

    handleAuthCancel() {
      this.cancelAuth();
      this.dialogClose();
    },

    async handleAccountRequest() {
      this.openCreateAccountPage();
    },

    async openCreateAccount() {
      const isExist = await this.checkAccountExists();
      if (!isExist) {
        this.activeForm = FORMS.CREATE_WALLET;
        await this.waitAccountCreate();
      }
    },

    async handleAuthorize({ serverMode } = {}) {
      if (this.showCreateAccount) {
        await this.openCreateAccount();
      }

      this.confirmAuth(serverMode);
    },
  },

  components: {
    Screen,
    VFrame,
    CompositeAuthForm,
    CreateWalletForm,
  },
};
</script>
