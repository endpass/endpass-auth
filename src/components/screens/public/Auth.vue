<template>
  <screen>
    <v-frame
      :loading="!isInited"
      :closable="false"
      @close="handleAuthCancel"
    >
      <composite-auth-form
        :closable="false"
        :is-public="true"
        @authorize="handleAuthorize"
      />
    </v-frame>
  </screen>
</template>

<script>
/* eslint-disable camelcase */

import { mapMutations, mapState, mapActions } from 'vuex';
import queryStringToMap from '@endpass/utils/queryStringToMap';
import Screen from '@/components/common/Screen';
import VFrame from '@/components/common/VFrame';
import CompositeAuthForm from '@/components/forms/CompositeAuth';
import { parseUrl } from '@/util/dom';

export default {
  name: 'PublicAuth',

  data: () => ({
    queryParamsMap: {},
  }),

  methods: {
    ...mapMutations(['setAuthParams']),
    ...mapState({
      isInited: state => state.core.isInited,
    }),
    ...mapActions(['cancelAuth', 'dialogClose']),

    handleAuthorize() {
      const { redirect_url } = this.queryParamsMap;

      if (redirect_url) {
        const fullPath = decodeURIComponent(redirect_url);

        const { origin } = parseUrl(fullPath);

        const newPath = fullPath.replace(origin, '');

        // const redirectRoute = decodeURIComponent(redirectUrl).replace(
        //   /^(https?:\/\/[a-z.]+(:\d+)?)/gm,
        //   '',
        // );

        this.$router.replace(newPath);
        // this.$router.replace(redirectRoute);
      }
    },

    handleAuthCancel() {
      this.cancelAuth();
      this.dialogClose();
    },
  },

  mounted() {
    const { search } = window.location;

    this.queryParamsMap = queryStringToMap(search);

    if (this.queryParamsMap.redirect_url) {
      this.setAuthParams({
        redirectUrl: decodeURIComponent(this.queryParamsMap.redirect_url),
      });
    }
  },

  components: {
    Screen,
    VFrame,
    CompositeAuthForm,
  },
};
</script>
