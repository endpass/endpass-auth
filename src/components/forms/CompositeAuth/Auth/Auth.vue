<template>
  <sign-in
    v-if="currentForm === FORMS.SIGN_IN"
    :is-public="isPublic"
    @social="onSocial"
    @submit="onSubmit"
    @switch="onSwitch"
  />
  <sign-up
    v-else-if="currentForm === FORMS.SIGN_UP"
    @social="onSocial"
    @submit="onSubmit"
    @switch="onSwitch"
  />
</template>

<script>
import SignIn from './SignIn';
import SignUp from './SignUp';
import { authStore } from '@/store';

const FORMS = {
  SIGN_IN: 'SIGN_IN',
  SIGN_UP: 'SIGN_UP',
};

export default {
  name: 'Auth',

  authStore,

  props: {
    isPublic: {
      type: Boolean,
      required: true,
    },
  },

  data: () => ({
    currentForm: FORMS.SIGN_IN,
    FORMS,
  }),

  methods: {
    onSwitch() {
      this.currentForm =
        this.currentForm === FORMS.SIGN_UP ? FORMS.SIGN_IN : FORMS.SIGN_UP;
    },

    async onSocial() {
      this.$emit('social');
    },

    async onSubmit(options) {
      this.$emit('submit', options);
    },
  },

  components: {
    SignIn,
    SignUp,
  },
};
</script>
