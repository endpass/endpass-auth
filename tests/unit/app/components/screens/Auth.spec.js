import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Auth from '@/components/screens/Auth.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Auth', () => {
  let store;
  let storeData;
  let wrapper;
  let coreModule;
  let accountsModule;

  beforeEach(() => {
    jest.clearAllMocks();

    coreModule = {
      state: {
        showCreateAccount: true,
      },
      getters: {
        isDialog: jest.fn(() => true),
      },
      actions: {
        dialogClose: jest.fn(),
      },
    };
    accountsModule = {
      actions: {
        cancelAuth: jest.fn(),
        confirmAuth: jest.fn(),
        checkAccountExists: jest.fn().mockResolvedValue(true),
        openCreateAccountPage: jest.fn(),
        waitAccountCreate: jest.fn(),
      },
    };
    storeData = {
      modules: {
        core: coreModule,
        accounts: accountsModule,
      },
    };
    store = new Vuex.Store(storeData);
    wrapper = shallowMount(Auth, {
      localVue,
      store,
    });
  });

  describe('render', () => {
    it('should correctly render Auth screen component', () => {
      expect(wrapper.find('composite-auth-form-stub').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render create account form if user authorized but does not have any account', async () => {
      expect.assertions(3);

      accountsModule.actions.checkAccountExists.mockResolvedValue(false);

      wrapper.find('composite-auth-form-stub').vm.$emit('authorize');

      await global.flushPromises();

      expect(wrapper.find('create-account-form-stub').exists()).toBe(true);
      expect(accountsModule.actions.waitAccountCreate).toBeCalled();
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render auth form if user authorized and have any account', async () => {
      expect.assertions(2);

      wrapper.find('composite-auth-form-stub').vm.$emit('authorize');

      await global.flushPromises();

      expect(wrapper.find('composite-auth-form-stub').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should confirm auth on auth form authorize event handling', async () => {
      expect.assertions(1);

      const payload = {
        serverMode: {
          foo: 'bar',
        },
      };

      wrapper.find('composite-auth-form-stub').vm.$emit('authorize', payload);

      await global.flushPromises();

      expect(accountsModule.actions.confirmAuth).toBeCalledWith(
        expect.any(Object),
        payload.serverMode,
        undefined,
      );
    });

    describe('create account form logic', () => {
      it('should open create account page on request event handling', () => {
        wrapper.setData({
          activeForm: 'CREATE_ACCOUNT',
        });

        wrapper.find('create-account-form-stub').vm.$emit('request');

        expect(accountsModule.actions.openCreateAccountPage).toBeCalled();
      });
    });
  });
});
