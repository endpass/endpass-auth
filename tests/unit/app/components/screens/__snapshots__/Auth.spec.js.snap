// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth render should correctly render Auth screen component 1`] = `
<screen-stub>
  <v-frame-stub title="Connect" closable="true" loading="true">
    <composite-auth-form-stub closable="true"></composite-auth-form-stub>
  </v-frame-stub>
</screen-stub>
`;

exports[`Auth render should render auth form if user authorized and have any account 1`] = `
<screen-stub>
  <v-frame-stub title="Connect" closable="true" loading="true">
    <composite-auth-form-stub closable="true"></composite-auth-form-stub>
  </v-frame-stub>
</screen-stub>
`;

exports[`Auth render should render create account form if user authorized but does not have any account 1`] = `
<screen-stub>
  <v-frame-stub title="Connect" closable="true" loading="true">
    <create-wallet-form-stub></create-wallet-form-stub>
  </v-frame-stub>
</screen-stub>
`;
